src/Service/ProducerEventProcessorTest.cpp

```diff
@@ -0,0 +1,1291 @@
+
+constexpr uint64_t kTimestampNs1 = 7723;
+constexpr uint64_t kTimestampNs2 = 7727;
+
+constexpr int32_t kNumBeginMarkers1 = 19;
+constexpr int32_t kNumBeginMarkers2 = 23;
+
+constexpr int32_t kDepth1 = 29;
+constexpr int32_t kDepth2 = 31;
+
+constexpr uint64_t kFunctionId1 = 37;
+constexpr uint64_t kFunctionId2 = 41;
+
+constexpr uint32_t kGpuJobContext1 = 43;
+constexpr uint32_t kGpuJobContext2 = 47;
+
+constexpr uint32_t kSeqNo1 = 53;
+constexpr uint32_t kSeqNo2 = 59;
+
+constexpr float kAlpha1 = 0.1f;
+constexpr float kRed1 = 0.2f;
+constexpr float kGreen1 = 0.3f;
+constexpr float kBlue1 = 0.4f;
+
+constexpr float kAlpha2 = 2.1f;
+constexpr float kRed2 = 2.2f;
+constexpr float kGreen2 = 2.3f;
+constexpr float kBlue2 = 2.4f;
+
+}  // namespace
+
+TEST(ProducerEventProcessor, OneSchedulingSliceEvent) {
+  MockCaptureEventBuffer buffer;
+  auto producer_event_processor = CreateProducerEventProcessor(&buffer);
+
+  ProducerCaptureEvent event;
+  SchedulingSlice* scheduling_slice = event.mutable_scheduling_slice();
+  scheduling_slice->set_pid(kPid1);
+  scheduling_slice->set_tid(kTid1);
+  scheduling_slice->set_core(kCore1);
+  scheduling_slice->set_duration_ns(kDurationNs1);
+  scheduling_slice->set_out_timestamp_ns(kTimestampNs1);
+
+  ClientCaptureEvent client_capture_event;
+  EXPECT_CALL(buffer, AddEvent).Times(1).WillOnce(SaveArg<0>(&client_capture_event));
+
+  producer_event_processor->ProcessEvent(kDefaultProducerId, event);
+  ASSERT_EQ(client_capture_event.event_case(), ClientCaptureEvent::kSchedulingSlice);
+  const SchedulingSlice& actual_scheduling_slice = client_capture_event.scheduling_slice();
+
+  EXPECT_EQ(actual_scheduling_slice.pid(), kPid1);
+  EXPECT_EQ(actual_scheduling_slice.tid(), kTid1);
+  EXPECT_EQ(actual_scheduling_slice.core(), kCore1);
+  EXPECT_EQ(actual_scheduling_slice.duration_ns(), kDurationNs1);
+  EXPECT_EQ(actual_scheduling_slice.out_timestamp_ns(), kTimestampNs1);
+}
+
+TEST(ProducerEventProcessor, OneInternedCallstack) {
+  MockCaptureEventBuffer buffer;
+  auto producer_event_processor = CreateProducerEventProcessor(&buffer);
+
+  ProducerCaptureEvent event;
+  InternedCallstack* interned_callstack = event.mutable_interned_callstack();
+  interned_callstack->set_key(kKey1);
+  Callstack* callstack = interned_callstack->mutable_intern();
+  callstack->add_pcs(1);
+  callstack->add_pcs(2);
+  callstack->add_pcs(3);
+
+  ClientCaptureEvent client_capture_event;
+  EXPECT_CALL(buffer, AddEvent).Times(1).WillOnce(SaveArg<0>(&client_capture_event));
+
+  producer_event_processor->ProcessEvent(kDefaultProducerId, event);
+
+  ASSERT_EQ(client_capture_event.event_case(), ClientCaptureEvent::kInternedCallstack);
+  const InternedCallstack& actual_interned_callstack = client_capture_event.interned_callstack();
+
+  // We do not expect resulting id to be the same, but we also do not
+  // enforce sequential ids in the test. 0 value is reserved.
+  EXPECT_NE(actual_interned_callstack.key(), 0);
```

Comment:
I think I mentioned somewhere else about adding a constant for the reserved 0. You could then use it in all these tests.

Reply:
done.